### ISSUE TITLE
Update of Graph with new nodes and edges

### ISSUE THREAD

**Issue by akshayagarwalgithub** [graphframes/graphframes#252]
How do we update a graph using Graphframe when new nodes and edges are added to it. When the new nodes and edges are added to the node and edge dataframes does the Graph needs to be build again or is there a way to optimize the update of graph using graphframes? I am looking for an efficient way of doing it.

**Comment by artem-aliev** [comment]
Graph frame itself has no update methods. It is for querying only.
It is a lightweight object (except cached vertex ids). It just references two DataFrames.
So if you add nodes and edges to underlying DataSource (say Database) they will be available for GF queries. But if you use Graph Algorithms (GraphX), vertex ids will be cached, and GrpahX will cache data frame also and you will not see your updates in that algorithms.It is fast to just create new GraphFrame in this case

**Comment by felixcheung** [comment]
correct... you could union the new edge DF to the existing and new vertex DF to the existing, and then pass the results to make a new GraphFrame; and then perform your graph operations again

**Comment by akshayagarwalgithub** [comment]
@felixcheung @artem-aliev  I have a question - So when I do union and create another dataframe, it will be an entirely new one. So when I run some algorithms on this new updated graphframe then would the computation be initiated from start(from time perspective)? I mean wouldn't Spark be re computing the Graph frame since the updated dataframes are entirely new dataframes(due to immutable feature of Spark)? I am looking forward to a way of optimizing so that recomputation of the previously computed dataframes does not happen while doing update to graphs.

**Comment by felixcheung** [comment]
that is true, computationally it will not be incremental - it will be rebuild from scratch.

**Comment by akshayagarwalgithub** [comment]
@felixcheung @artem-aliev  Two questions here:
1. Would'nt this be a very big challenge, where a Big graph with billions of vertices and edges has to be recomputed again every day for computation when a small incremental data load happens to the nodes and edges dataframe?
2. If the node or edge dataframe is pointing to a table in an external database by JDBC connection, and say there is an update in that table in external database. Would the computation on Graph still be done from scratch, when we are not doing any union or update logic on dataframes in the Spark application ourselves, but it is happening in the source database?

**Comment by akshayagarwalgithub** [comment]
@felixcheung @ankurdave Looking forward to your response on this.

**Comment by harryyy213** [comment]
1. If you are using Apache spark with HDFS for data storage, one thing you need to understand is HDFS is not a native graph database like Neo4j so it can't store a graph, what it stores is information of vertices and edges as data tables. Apache spark just provides a Graph API using which graph can be build from the vertices and edges table stored in HDFS. Said that, what it means is, a graph as such can't be stored rather data tables (vertices and edges) are stored and graph is constructed on the fly by graph API. Once this is clear you will understand the challenge you are mentioning is true but unavoidable when working with graphframes. However, the time taken to construct a graph is really very quick and new addition of data is not going to drastically increase you time. 

2. Yes, if the nodes and edges tables are being taken from an external database and then being directly used in graphframes and if these tables get updated the computation of graph will happen form the scratch, given:
a.  There is an automatic or triggered update procedure set-up for the graphframes to re-run whenever change in external database happens for graph computation to happen. 
b.  Apache Spark is a lazy evaluation procedure, which means  the execution of graphframes (construction of graph) will not happen unless and until some statement needs to use graph for its execution (e.g.: any algorithm or display statements)
If none of the above procedures are happening and the graph is already created. Any changes in external database will not be reflected in you graph.

**Comment by SemyonSinchenko** [comment]
I do not think resolving this issue is feasible anyhow. All the algorithms are written in a way to work from scratch, there is no easy way to change it.
